### ISSUE TITLE
Would this library support generating namespaced UUIDs without a namespace?

### ISSUE THREAD

**Issue by alexanderankin** [uuidjs/uuid#532]
## Is your feature request related to a problem? Please describe.

I am working with some other microservices implemented in Java, their acceptance tests use UUIDs which are generated from string names, and no namespaces, using this library: [docs](http://cowtowncoder.github.io/java-uuid-generator/javadoc/3.1.3/index.html?com/fasterxml/uuid/impl/NameBasedGenerator.html), [source](https://github.com/cowtowncoder/java-uuid-generator/blob/master/src/main/java/com/fasterxml/uuid/impl/NameBasedGenerator.java#L136).

## Describe the solution you'd like

I think it would be possible to include a similar option, which omits the namespace if it is not given.

## Describe alternatives you've considered

I am considering just using the URL namespace, but i was hoping to raise as little eyebrows as possible for future workers who have to read my code. Tests are like documentation after all, and I was hoping to reflect the logic as closely as possible as the Java side.

It is also possible to give `{ length: 16 }` as the argument instead of a uuid string, which achieves the desired result, but would be preferable to just give null as the argument, or something along those lines.

## Additional context

Here is the logic which i was hoping to get from this library, in Java code (please forgive me, i couldnt get a job doing nodejs in my city):

```java
    @Override
    public UUID generate(byte[] nameBytes)
    {
        byte[] digest;
        synchronized (_digester) {
            _digester.reset();
            if (_namespace != null) {
                _digester.update(UUIDUtil.asByteArray(_namespace));
            }
            _digester.update(nameBytes);
            digest = _digester.digest();
        }
        return UUIDUtil.constructUUID(_type, digest);
    }
```

**Comment by alexanderankin** [comment]
actually, it looks like i can fool it with `{ length: 16 }` instead of a string. but i figured i would let you guys know what is being implemented out there. it works in that i get the same result every time i do `uuid.v5('name', { length: 16 })`

**Comment by broofa** [comment]
The way I read RFC4122, it doesn't really allow for this.  [The namespace algorithm](https://tools.ietf.org/html/rfc4122#section-4.3) states:
* **Allocate a UUID to use as a "name space ID" for all UUID generated from names in that name space**; see Appendix C for some pre-defined values.
* ...
* Compute the hash of the **name space ID** concatenated with the name.

The Java code you reference, if I'm reading it correctly, omits the "concatenated with" step if no namespace is provided.   I.e. it just hashes on the name and nothing else.  While I get that an argument could be made for why that is useful(???), I see it as a bit of a liability.  E.g. if someone were to inadvertantly pass a null or undefined namespace, it wouldn't error out and would lead to problems (potentially significant problems) later on.

Regarding your `{length: 16}` workaround, what you're effectively doing is passing the NIL id as the namespace.  I.e.:
```
const {NIL, v5} = require('uuid');
console.log(v5('hi', {length: 16}) === v5('hi', NIL))  // => true
```

This is a perfectly acceptable use case, by the way.  It's just not the same thing as not using the namespace at all.  So... do what you will with that information.

**Comment by alexanderankin** [comment]
Oh, that's interesting, thanks for that insight. Yes, that is what i think that code is doing also.

**Comment by alexanderankin** [comment]
it seems as though it is a different result in the other implementation:

```java
import java.util.UUID;
import com.fasterxml.uuid.Generators;
public class Main {
    public static void main(String[] args) {
        System.out.println("uuid"+ Generators.nameBasedGenerator().generate("hello").toString());
        System.out.println("uuid"+ Generators.nameBasedGenerator(new UUID(0, 0)).generate("hello").toString());
    }
}
```

produces:

```
uuidaaf4c61d-dcc5-58a2-9abe-de0f3b482cd9
uuidb7502f40-1152-59f2-ba10-69aeed522cdf
```

whereas:

```js
var uuid = require('uuid');
console.log('uuid' + uuid.v5('hello', uuid.NIL))
```

produces

```
uuidb7502f40-1152-59f2-ba10-69aeed522cdf
```

thanks for the quick response and the tip once again

**Comment by ctavan** [comment]
@alexanderankin yes, this is totally expected. The two different UUIDs you are generating with the JAVA library are:

```
hash(name) != hash(NIL + name)
```

where `NIL` are just 16 0-bytes.

**Comment by Alan-Liang** [comment]
I came across this when trying to implement [the Yggdrasil protocol for Minecraft](https://github.com/yushijinhun/authlib-injector/wiki/Yggdrasil-%E6%9C%8D%E5%8A%A1%E7%AB%AF%E6%8A%80%E6%9C%AF%E8%A7%84%E8%8C%83#%E5%85%BC%E5%AE%B9%E7%A6%BB%E7%BA%BF%E9%AA%8C%E8%AF%81); if you **really** need to generate the same UUID as `UUID.nameUUIDFromBytes(bytes)` in Java does, here is a hack to trick the library into doing this:

```js
const uuid = require('uuid')
const str = 'hello'
uuid.v5(
  {
    ...[ ...str ].map(x => x .charCodeAt(0)),
    get length () {
      return this.ok
        ? str.length
        : (this.ok = 1, str.length - 16)
    },
  },
  {
    get length () {
      return this.ok
        ? 0
        : (this.ok = 1, 16)
    },
  },
) // => 'aaf4c61d-dcc5-58a2-9abe-de0f3b482cd9'
```

It's a hack anyway (so pin the npm package version), and I really hope we can achieve this in an easier way.

**Comment by gobengo** [comment]
I don't believe that that https://datatracker.ietf.org/doc/html/rfc9562 requires namespace ids to have a particular length, but this package currently requires that it be exactly 16 bytes long.

This fork modifies the v35 (v3 and v5 uuid) function to allow any bytes as namespace id, which I believe is valid per RFC9562
https://github.com/gobengo/uuid

**Comment by broofa** [comment]
@gobengo To my knowledge, it was never the intent of RFC9562 authors to change the requirements around namespace structure.  There was [no discussion](https://github.com/search?q=repo%3Auuid6%2Fuuid6-ietf-draft+namespace&type=issues) about doing so when the new spec was authored.  And the term "Namespace ID" is used liberally throughout, which I would interpret as meaning the intent is still there, even if the verbiage that explicitly required it was inadvertently removed.

tl;dr: If non-UUID namespaces are no longer required, that's almost certainly a regression that I would expect to see addressed in a future version of the spec.
